File: All Languages/Java Codes/InsertionSort.java
package algorithm;

public class InsertionSort<T extends Comparable<T>> {
    private T[] a;

    public InsertionSort(T[] b) {
        a = b;
    }

    public InsertionSort() {

    }

    public T[] sort() {
        return sort(a);
    }

    public T[] sort(T[] a) {
        for(int i=1;i<a.length;i++) {
            int j = i-1;
            T key = a[i];

            while((j >= 0) && (a[j].compareTo(key) > 0)) {
                a[j+1] = a[j];
                j--;
            }
            a[j+1] = key;
        }
        return a;
    }
}

File: All Algorithms/insertion_sort.cpp
// C++ program for insertion sort
 
#include <bits/stdc++.h>
using namespace std;
 
// Function to sort an array using insertion sort
void insertionSort(int arr[], int n)
{
    int i, key, j;
    for (i = 1; i < n; i++)
    {
        key = arr[i];
        j = i - 1;
        while (j >= 0 && arr[j] > key)
        {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
    }
}
 
// A utility function to print an array of size n
void printArray(int arr[], int n)
{
  int i;
    for (i = 0; i < n; i++)
        cout << arr[i] << " ";
    cout << endl;
}

int main()
{
    int arr[] = { 10, 11, 15, 25, 4 };
    int N = sizeof(arr) / sizeof(arr[0]);
 
    insertionSort(arr, N);
    printArray(arr, N);
 
    return 0;
}
